runtime/lua/vim/lsp/folding_range.lua

```diff
@@ -0,0 +1,417 @@
+        end
+      end
+    end,
+  })
+  api.nvim_create_autocmd('LspAttach', {
+    group = augroup_setup,
+    buffer = bufnr,
+    callback = function(args)
+      local client = assert(vim.lsp.get_client_by_id(args.data.client_id))
+      request(bufnr, client)
+    end,
+  })
+  api.nvim_create_autocmd('LspNotify', {
+    group = augroup_setup,
+    buffer = bufnr,
+    callback = function(args)
+      local client = assert(vim.lsp.get_client_by_id(args.data.client_id))
+      if
+        client:supports_method(ms.textDocument_foldingRange, bufnr)
+        and (
+          args.data.method == ms.textDocument_didChange
+          or args.data.method == ms.textDocument_didOpen
+        )
+      then
+        request(bufnr, client)
+      end
+    end,
+  })
+
+  for _, client in
+    ipairs(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange }))
+  do
+    request(bufnr, client)
+  end
+
+  return bufstates[bufnr]
+end
+
+-- The "garbage collector" for `bufstate` and its associated event hooks.
+-- See the comment above `augroup_setup` function for more details.
+--
+-- It will be fine without this, as the bufstate and event hooks will also
+-- be cleared when the buffer is unloaded.
+-- api.nvim_create_autocmd('OptionSet', {
+--   group = augroup_setup,
+--   pattern = 'foldexpr',
+--   callback = function()
+--     -- Sould find a more reliable way to detect changes in `foldexpr`.
+--     if vim.v.option_old == 'v:lua.vim.lsp.foldexpr()' then
+--       for bufnr in pairs(bufstates) do
+--         if
+--           ---@param winid integer
+--           vim.iter(vim.fn.win_findbuf(bufnr)):all(function(winid)
+--             return vim.wo[winid].foldexpr ~= 'v:lua.vim.lsp.foldexpr()'
+--           end)
+--         then
+--           bufstates[bufnr] = nil
+--           api.nvim_clear_autocmds({ buffer = bufnr, group = augroup_setup })
+--         end
+--       end
+--     end
+--   end,
+-- })
+
+---@param kind lsp.FoldingRangeKind
+---@param winid integer
+local function foldclose(kind, winid)
+  vim._with({ win = winid }, function()
+    local bufnr = api.nvim_win_get_buf(winid)
+    local row_kinds = bufstates[bufnr].row_kinds
+    -- Reverse traverse to ensure that the smallest ranges are closed first.
+    for row = api.nvim_buf_line_count(bufnr) - 1, 0, -1 do
+      local kinds = row_kinds[row]
+      if kinds and kinds[kind] then
+        vim.cmd(row + 1 .. 'foldclose')
+      end
+    end
+  end)
+end
+
+--- Close all {kind} of folds in the the window with {winid}.
+---
+--- To automatically fold imports when opening a file, you can use an autocmd:
+---
+--- ```lua
+--- vim.api.nvim_create_autocmd('LspNotify', {
+---   callback = function(args)
+---     if args.data.method == 'textDocument/didOpen' then
+---       vim.lsp.folding_range.foldclose('imports', vim.fn.bufwinid(args.buf))
+---     end
+---   end,
+--- })
+--- ```
+---
+---@param kind lsp.FoldingRangeKind "comment" | "imports" | "region"
```

Comment:
For consistency with other docs:

```suggestion
---@param kind lsp.FoldingRangeKind 'comment'|'imports'|'region'
```

Reply:
It seems that changing it to the form of `'` will cause lintdoc to error, possibly because it is confused with type unions.

Reply:
Oh the type is `lsp.FoldRangeKind`. These can be fully omitted.

Reply:
I slightly added content, keeping it in single quotes instead of double quotes.

Reply:
> Oh the type is `lsp.FoldRangeKind`. These can be fully omitted.

The purpose of this comment is to generate documentation, as `lsp.FoldingRangeKind` will not be automatically generated, so I mentioned their possible values.